torch/csrc/Module.cpp

```diff
@@ -688,6 +690,38 @@ PyObject* THPModule_isEnabledXNNPACK(PyObject* _unused, PyObject* noargs) {
     Py_RETURN_FALSE;
 }
 
+PyObject* THPModule_willEngineExecuteNode(PyObject* _unused, PyObject* arg) {
+  HANDLE_TH_ERRORS
+  bool isTHPFunction = THPFunction_Check(arg);
+  bool isTHPCppFunction = torch::autograd::THPCppFunction_Check(arg);
+  THPUtils_assert(
+      isTHPFunction || isTHPCppFunction,
+      "_will_engine_execute_node expects an grad_fn, "
+      "but got %s",
+      THPUtils_typename(arg));
+  const auto exec_info = torch::autograd::get_current_graph_task_exec_info();
+  THPUtils_assert(
+      exec_info,
+      "_get_should_execute_nodes should only be called during the backward pass");
+  THPUtils_assert(
+      !exec_info->empty(),
+      "_get_should_execute_nodes should only be used with .grad() or .backward() "
+      "if the `inputs` is passed.");
+  torch::autograd::Node* node;
+  if (isTHPFunction) {
+    node = ((THPFunction*)arg)->cdata.lock().get();
```

Comment:
Is this OK? Aren't you obligated to keep lock's output live?

Reply:
Good catch, updated

Reply:
actually since we don't dereference the ptr it is probably OK, though it makes sense in principle

Reply:
Keeping it live prevents ABA races, although since you take out a mutex that might also be sufficient protection